### PR TITLE
🎉 Source Talkdesk: add region support

### LINKED ISSUE #17567: Source TalkDesk: add support for Europe and Canada regions
## Environment
- **Airbyte version**:  0.40.3
- **OS Version / Instance**:   AWS EC2
- **Deployment**: Docker 
- **Source Connector and version**: TalkDesk 
- **Destination Connector and version**: Snowflake 
- **Step where error happened**: Sync job

## Current Behavior
When trying to sync data from TalkDesk  which is hosted on a `.eu` domain we get an HTTP401 error because it tries to access the api on a `.com` domain:

```requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://api.talkdeskapp.com/data/reports/calls/jobs```

For context see [this](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/streams.py#L35).

## Expected Behavior
Sync should work fine. For this to occur the API url should be on the `.eu` domain: `https://api.talkdeskapp.eu/data/reports/calls/jobs`

## Logs

[logs-154.txt](https://github.com/airbytehq/airbyte/files/9708796/logs-154.txt)

## Steps to Reproduce
1. Set up eu based TalkDesk source
2. Try to sync data into a destination
3. Get 401 error

## Are you willing to submit a PR?

If we do some workaround internally in our team we could try to share it, but i'm not sure if we will do a complete solution or replace the hardcoded url.

### PR DESCRIPTION
**Pull request by sh4sh**
## What
Implements https://github.com/airbytehq/airbyte/issues/17567

Talkdesk Explore supports [US, Canada and Europe regions](https://docs.talkdesk.com/docs/how-to-guarantee-your-app-works-in-all-regions#supported-regions-and-base-urls), but currently this connector only supports the US region.

## How
Since all three regions offer the same endpoints just with a different URL, we can add a connector config option to specify the desired region and build the URL base property to match.

## Recommended reading order
1. `spec.json`
2. `source.py`
3. `streams.py`

## 🚨 User Impact 🚨
Users should now be able to configure which Talkdesk region they want to use.

Screen Shot 2022-10-05 at 9 08 41 AM

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

```
% python -m pytest -s unit_tests
Test session starts (platform: darwin, Python 3.10.6, pytest 6.2.5, pytest-sugar 0.9.5)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/sunny/temporary/airbyte/airbyte-integrations/connectors/source-talkdesk-explore/.hypothesis/examples')
rootdir: /Users/sunny/temporary/airbyte, configfile: pytest.ini
plugins: hypothesis-6.54.6, sugar-0.9.5, requests-mock-1.9.3, mock-3.6.1, timeout-1.4.2, cov-3.0.0
collecting ...
 airbyte-integrations/connectors/source-talkdesk-explore/unit_tests/unit_test.py::test_example_method ✓       100% ██████████

Results (1.31s):
       1 passed
```

</details>

<details><summary><strong>Integration</strong></summary>

```
% python -m pytest -s integration_tests
Test session starts (platform: darwin, Python 3.10.6, pytest 6.2.5, pytest-sugar 0.9.5)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/sunny/temporary/airbyte/airbyte-integrations/connectors/source-talkdesk-explore/.hypothesis/examples')
rootdir: /Users/sunny/temporary/airbyte, configfile: pytest.ini
plugins: hypothesis-6.54.6, sugar-0.9.5, requests-mock-1.9.3, mock-3.6.1, timeout-1.4.2, cov-3.0.0
collecting ...

Results (0.08s):
```

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by sh4sh** [comment]
This PR has not been tested as we are still waiting on a Talkdesk sandbox account in https://github.com/airbytehq/airbyte/issues/10285

**Comment by sh4sh** [comment]
@ggmblr Hmm, in your screenshot it looks like you've selected the official image, this PR update will not be reflected there until we're able to merge and publish. I do see `Region` when I build it as a custom image though, I pushed it to dockerhub and followed [these instructions](https://docs.airbyte.com/integrations/custom-connectors#adding-your-connectors-in-the-ui) to add it into the UI:

Screen Shot 2022-10-05 at 9 08 41 AM

**Comment by ggmblr** [comment]
Ah sorry about that. Which are the steps that i need to follow then? Currently i'm on the PR branch and just executed the docker-compose up. How do i use the custom image?

**Comment by sh4sh** [comment]
It's all good, thanks for checking this out for me!

Normally you'd have to build and push the image to dockerhub or another image repository so that Airbyte can pull it. However, you can also just use mine: https://hub.docker.com/repository/docker/sh4sh/source-talkdesk-explore

Then in the Airbyte UI, go to `Settings` > `Sources` > `+ New connector`.
You'll see the screen below, and you should add these details:

Connector display name: `Talkdesk Explore TEST` (this value doesn't really matter, it's just the name that displays in the UI)
Docker repository name: `sh4sh/source-talkdesk-explore`
Docker image tag: `dev`
Connector Documentation URL: `https://docs.airbyte.com/integrations/sources/talkdesk-explore`

Then hit `Add`, go to Sources and add the connector normally. Hopefully you're able to trigger a sync! But I haven't tested that part yet...

Screen Shot 2022-10-05 at 9 58 33 AM

**Comment by ggmblr** [comment]
@sh4sh Thanks for the explanation. I got it up, although it did fail in the sync, something about missing the region argument in the __init__ in streams.py:

```
2022-10-05 14:19:21 [44msource[0m > Syncing stream: calls 
2022-10-05 14:19:21 [44msource[0m > Encountered an exception while reading stream calls
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 119, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 159, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 215, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 356, in read_records
    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
  File "/airbyte/integration_code/source_talkdesk_explore/streams.py", line 109, in path
    base_path=self.base_path, start_date=latest_state, timezone=self.timezone, authenticator=self.authenticator
TypeError: __init__() missing 1 required positional argument: 'region'
2022-10-05 14:19:21 [44msource[0m > Finished syncing calls
2022-10-05 14:19:21 [44msource[0m > SourceTalkdeskExplore runtimes:
Syncing stream calls 0:00:00.074147
2022-10-05 14:19:21 [44msource[0m > __init__() missing 1 required positional argument: 'region'
Traceback (most recent call last):
  File "/airbyte/integration_code/main.py", line 13, in <module>
    launch(source, sys.argv[1:])
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 127, in launch
    for message in source_entrypoint.run(parsed_args):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 118, in run
    for message in generator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 123, in read
    raise e
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 119, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 159, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 215, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 356, in read_records
    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
  File "/airbyte/integration_code/source_talkdesk_explore/streams.py", line 109, in path
    base_path=self.base_path, start_date=latest_state, timezone=self.timezone, authenticator=self.authenticator
TypeError: __init__() missing 1 required positional argument: 'region'
```

Full logs:
[talk_desk_pr_logs.txt](https://github.com/airbytehq/airbyte/files/9716422/talk_desk_pr_logs.txt)

**Comment by sh4sh** [comment]
@ggmblr Oops, should be fixed now! I've uploaded a new image under the same name and tag, so if you create a new source connector the same way as before then it should hopefully pull the new image.

**Comment by ggmblr** [comment]
@sh4sh I tried simply clicking change on the source connector, and i tried to create a new one with the same config, but both of them failed with the same error as before (?)

```
2022-10-06 07:06:01 [44msource[0m > Syncing stream: calls 
2022-10-06 07:06:02 [44msource[0m > Encountered an exception while reading stream calls
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 119, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 159, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 215, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 356, in read_records
    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
  File "/airbyte/integration_code/source_talkdesk_explore/streams.py", line 109, in path
    base_path=self.base_path, start_date=latest_state, timezone=self.timezone, authenticator=self.authenticator
TypeError: __init__() missing 1 required positional argument: 'region'
2022-10-06 07:06:02 [44msource[0m > Finished syncing calls
2022-10-06 07:06:02 [44msource[0m > SourceTalkdeskExplore runtimes:
Syncing stream calls 0:00:00.087588
2022-10-06 07:06:02 [44msource[0m > __init__() missing 1 required positional argument: 'region'
Traceback (most recent call last):
  File "/airbyte/integration_code/main.py", line 13, in <module>
    launch(source, sys.argv[1:])
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 127, in launch
    for message in source_entrypoint.run(parsed_args):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 118, in run
    for message in generator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 123, in read
    raise e
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 119, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 159, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 215, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 356, in read_records
    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
  File "/airbyte/integration_code/source_talkdesk_explore/streams.py", line 109, in path
    base_path=self.base_path, start_date=latest_state, timezone=self.timezone, authenticator=self.authenticator
TypeError: __init__() missing 1 required positional argument: 'region'
```

Full logs:
[full_logs.txt](https://github.com/airbytehq/airbyte/files/9722105/full_logs.txt)

**Comment by ggmblr** [comment]
Hello @sh4sh !
I tried again with the new tag and it worked fine!

I'll leave here the logs as proof. When do you expect this to be merged?


[success_logs.txt](https://github.com/airbytehq/airbyte/files/9731423/success_logs.txt)

**Comment by ggmblr** [comment]
Hi @sh4sh !
Any update as to when we could expect this in production?

**Comment by sh4sh** [comment]
Hey @ggmblr sorry for the delay here! We are currently waiting on credentials so that our CI can run the acceptance tests... https://github.com/airbytehq/airbyte/issues/10285

However, if you are willing to run the tests with your credentials that would speed things up! We have instructions in the docs here: https://docs.airbyte.com/contributing-to-airbyte/developing-locally#run-acceptance-tests
If you're able to, please post the output here and we will publish the connector :)

**Comment by alejobs** [comment]
hi team, I am also interested in this PR. When do you expect to have the credentials? 

Thanks!

**Comment by sh4sh** [comment]
Hey @alejobs , it looks like we may have news this week but it's not a for sure thing. It's been very difficult to obtain a sandbox account for this particular service. 

If you have credentials for Talkdesk, it would be very helpful if you could run the acceptance tests! We would publish this connector if you could post a screenshot of the test output. We have instructions in the docs here: https://docs.airbyte.com/contributing-to-airbyte/developing-locally#run-acceptance-tests

**Comment by sh4sh** [comment]
Hey @ggmblr @alejobs , are either of you a current talkdesk customer? If so, we could use your help in getting sandbox credentials so this connector can be tested with our CI.

Please check this thread and let @igrankova know if you can help: https://github.com/airbytehq/airbyte/issues/10285#issuecomment-1298197802

>@sh4sh, hi!
We are stuck with the sandbox, no answer again. If you are a Talkdesk customer, could you help us and contact your Talkdesk sales manager and ask them about possibility to provide us a sandbox? Maybe they could agree to contact me ([iryna.grankova@airbyte.io](mailto:iryna.grankova@airbyte.io)). Usually it works very good if we have a recommendation from the mutual customer.

**Comment by ggmblr** [comment]
Hi @sh4sh ,
We talked to our team that is in contact with talkdesk, they have now opened a ticket to their support. Let's see if we can help you push on this.

File: airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/spec.json
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/spec.json:33]: You can use `enum` to have a better user experience, check other connectors spec. I don't have one example but let me know if you need help here.

File: airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/streams.py
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/streams.py:39]: I would suggest to create a constant outside the class
```python
MAP_REGION_DOMAIN = {"US": ".com"...}
```

I think this source can be refactored. Create a BaseStream(HttpStream) where only overwrite the url_base method and make both classes inherent from the BaseStream. With this you don't duplicate code.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Maybe would be nice refactor the code to achieve better readability and remove duplicate code.

File: airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/streams.py
Comment (sh4sh) [reply airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/streams.py]: Done here: https://github.com/airbytehq/airbyte/pull/17587/commits/93ba88ffb4fa74c1f02cf46f61f912f67ed675ba
hopefully it looks better let me know!

File: airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/spec.json
Comment (sh4sh) [reply airbyte-integrations/connectors/source-talkdesk-explore/source_talkdesk_explore/spec.json]: Done in https://github.com/airbytehq/airbyte/pull/17587/commits/a7df72dc59e2e69f38eb3d21c15e71e48e07e57e and https://github.com/airbytehq/airbyte/pull/17587/commits/731df211ca966b64751cd50897a284afe9c7dec6

**Comment by ggmblr** [comment]
Hi @sh4sh we got an answer from TalkDesk.  Here it is:

-----------
Would you kindly instruct the Airbyte team to follow [this guide](https://docs.talkdesk.com/docs/becoming-a-talkdesk-partner-developer) to become a Talkdesk Partner Developer. The Talkdesk Support team is not equipped to assist with this process.
Any queries regarding the process can be made to this email address: [appconnect@talkdesk.com](mailto:appconnect@talkdesk.com)
If questions arise regarding APIs, scopes, and other technical subjects, they can reach out to us through the [support web form](https://support.talkdesk.com/hc/en-us/requests/new).

-----------

I'm not sure if this is helpful for you guys, but is the only thing our related department could get.

**Comment by ggmblr** [comment]
Hi @sh4sh 

After some internal discussions we decided that we will try to help you with testing the PR.

Could you please let me know if the PR is ready to be tested?

Also could you maybe walk me through what are the steps that i need to follow? I feel the documentation you linked is taking some steps for granted.

**Comment by sh4sh** [comment]
Hey @ggmblr thanks so much for following up! We are in the partner onboarding process, it's taking a while but this should at least be resolved for any future PRs: https://github.com/airbytehq/airbyte/issues/10285#issuecomment-1328966249

In the meantime, if you're willing to run the acceptance tests that would be incredibly helpful! You would need to run [the tests in this doc](https://docs.airbyte.com/contributing-to-airbyte/developing-locally#run-acceptance-tests) on the `sh4sh/17567-talkdesk-region-support` branch. They may take some time to complete.
```
SUB_BUILD=PLATFORM ./gradlew clean build
SUB_BUILD=PLATFORM ./gradlew :airbyte-tests:acceptanceTests
``` 
I've just rebased the branch so it should be ready to go by the time you see this message. Feel free to reach out here or in Slack if I can help. Thanks again!

**Comment by alejobs** [comment]
Hi @sh4sh , I see this is still not merged, but in #10285 it seems it is unblocked. Can you please update me on this?

Thanks

cc: @igrankova

**Comment by igrankova** [comment]
hi, @sh4sh,
We already have an account for Talkdesk, [details](https://github.com/airbytehq/airbyte/issues/10285#issuecomment-1411879258). Is there any other blocker for merging this PR? 
cc @alejobs

**Comment by alejobs** [comment]
Hi @sh4sh , any updates on this? thanks

**Comment by ggmblr** [comment]
@sh4sh @igrankova any news on when this PR can be merged?

**Comment by alejobs** [comment]
hi @sajarin, @igrankova,

This PR is blocked since months ago and supposedly the blocker (talkdesk sandbox account) was already solved long time ago. Can you please help us merge it?

Thanks.

**Pull request closed without merging**
